core/src/main/scala/org/allenai/common/SeekableSource.scala

```diff
@@ -0,0 +1,256 @@
+
+  /** Reads and returns a single character from a the file. */
+  override def next(): Char = {
+    if (inBuffer.hasRemaining || (inputRemaining && fillBuffer())) {
+      if (useUtf8) {
+        nextUtf8
+      } else {
+        nextIso8859
+      }
+    } else {
+      throw new NoSuchElementException("next() called on empty SeekableSource")
+    }
+  }
+
+  /** Positions this source at the given byte offset in the input file. Note that this will allow
+    * you to position past the end of the file, in which case no input will be read.
+    */
+  def position(newPosition: Long): Unit = {
+    inFile.position(newPosition)
+    inBuffer.limit(0)
+    inputRemaining = true
+  }
+
+  /** @return the current position in the file */
+  def position: Long = inFile.position - inBuffer.remaining
+
+  /** @return the next character in the buffer, decoded from ISO-8859-1 */
+  protected[common] def nextIso8859: Char = {
+    // `toChar` doesn't work as expected with bytes; they're first widened with sign extension,
+    // meaning that bytes with the high bit set expand into two-byte characters with the high byte
+    // set to 0xff. Mask into an int first to fix this.
+    (inBuffer.get() & 0xff).toChar
+  }
+
+  /** @return the next character in the buffer, decoded from UTF-8 */
+  protected[common] def nextUtf8: Char = {
+    // Ensure we are either at the end of file, or have at least three bytes that are readable.
+    if (inBuffer.remaining() < 3 && inputRemaining) {
+      fillBuffer()
+    }
+    // See https://en.wikipedia.org/wiki/UTF-8 for details. This is checking the first byte for the
+    // total length of the encoding.
+    val first = inBuffer.get()
+    val intVal = if ((first & 0x80) == 0) {
+      return first.toChar
+    } else if ((first & 0xe0) == 0xc0) {
+      // TODO(jkinkead): Secondary bytes (here and below) should all have their two highest bits set
+      // as 10. This should be asserted somehow, and if they're corrupted, producing an unknown
+      // character instead.
+      val second = inBuffer.get()
+      ((first & 0x1f) << 6) | (second & 0x3f)
+    } else if ((first & 0xf0) == 0xe0) {
+      val second = inBuffer.get()
+      val third = inBuffer.get()
+      ((first & 0x0f) << 12) | ((second & 0x3f) << 6) | (third & 0x3f)
+    } else {
+      // Anything here is technically illegal UTF-8. However, some sources (i.e. Wikipedia) use
+      // wide UTF-8 encodings. This will attempt to skip the correct number of bytes. Note that
+      // while we could decode the character here, it will be wider than 16 bits, and therefore
+      // won't fit in a JVM Char.
+
+      // skipCount stores the *additional* bytes we will skip, not the total length of the encoding.
+      val skipCount = if ((first & 0xf8) == 0xf0) {
+        3
+      } else if ((first & 0xfc) == 0xf8) {
+        4
+      } else if ((first & 0xfe) == 0xfc) {
+        5
+      } else {
```

Comment:
I read somewhere that UTF-8 was designed so you can pick up a stream in the middle. There is some way to figure out where the next character starts. If that's true, you shouldn't have to do this. On the other hand, we're in the realm of illegal UTF-8, so maybe that doesn't apply here anyways.